index.js

```diff
@@ -421,7 +436,22 @@ function Physics (mcData, world) {
 
   function isOnLadder (world, pos) {
     const block = world.getBlock(pos)
-    return (block && (block.type === ladderId || block.type === vineId))
+    if (!block) { return false }
+    if (block.type === ladderId || block.type === vineId) { return true }
+
+    // Since 1.9, when a trapdoor satisfies the following conditions, it also becomes climbable:
+    //  1. The trapdoor is placed directly above a ladder.
+    //  2. The trapdoor is opened.
+    //  3. The trapdoor and the ladder directly below it face the same direction.
+    if (supportFeature('climbableTrapdoor') && trapdoorIds.has(block.type)) {
+      const blockBelow = world.getBlock(pos.offset(0, -1, 0))
+      if (blockBelow.type !== ladderId) { return false } // condition 1.
+      if (!block.getProperties().open) { return false } // condition 2.
+      if (block.getProperties().facing !== blockBelow.getProperties().facing) { return false } // condition 3
+      return true
+    }
```

Comment:
These are expensive calls, and hard to optimize for the JIT compiler as there's many allocations here. For physics code that needs to be run at high performance, it's not good. There should be a more optimal way to do this.

Reply:
Sorry I don't understand. Is that `.getProperties()` the expensive call, or is it `world.getBlock()`?

Reply:
This function is hot code, as it's run on every movement. Especially for things like pathfinding, it will be a bottleneck as several temporary allocations are done and end up as garbage.

Calling check feature is slow and can require iterating over all the features, so that instead can be a top level boolean. The 3 calls getProperties to all create copies of property data with property access. That can be done once.

Reply:
You are right, but this is partly due to both `supportFeature` and `.getProperties()` is badly designed and implemented. `supportFeature` should have used numeric ids instead of strings and `.getProperties()` should always return the same object (pointer).

Reply:
p.s. This branch is defiantly **not** hot code. In practice, the bot will not frequently step into trapdoors

Reply:
It is hot, and isOnLadder is called twice on each movement tick. Also, all the code inside a hot function, whether it's behind a branch or not is factored into whether or not a function can be optimized or not.

https://github.com/PrismarineJS/prismarine-physics/blob/d29774ae4ca426d87a5209a9dc5e92aa45458b7c/index.js#L580

Reply:
Ideally we also look into if getBlock() calls themselves can be faster, perhaps by using state IDs, but that's beyond the scope of the PR

Reply:
https://github.com/PrismarineJS/prismarine-block/blob/38e8e877e43c1ad29a90d23102a6c23e0d95ebbd/index.js#L263

Would adding a new function to prismarine-block help with this?
getPropertiesPointer () {
    //whichever of these would help solve this ladder issue
    return this._properties;
    return this.computedStates;
}


Alternatively making the feature passable as a boolean so that things simulating physics need not worry about this unless they toggle it on, but applied movement would still work fine. Access to the feature, even in a cumbersome manner with room for optimization, will still give people a solution.